File: Python/deleteDuplicates.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        cur= head
        while cur and cur.next:
            if cur.val == cur.next.val:
                cur.next = cur.next.next
            else:
                cur = cur.next
        return head

        # method 2 use dict
        dic = {}
        cur = head
        while cur:
            dic[cur.val] = dic.get(cur.val, 0) + 1
            cur = cur.next
        cur = head
        while cur:
            tmp = cur
            for _ in xrange(dic[cur.val]):
                tmp = tmp.next
            cur.next = tmp
            cur = cur.next
        return head
        
        

File: src/leetcode/S0083_remove-duplicates-from-sorted-list/deleteDuplicates.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
// [1,1,1] => [1]
var deleteDuplicates = function (head) {
  let node = head;
  while (node && node.next) {
    // 一定要注意这里的 else，如果值相等，只是先改变 node 的 next 联接，但是 node 指针还暂时不往后走
    // 没有 else 包裹会导致 [1,1,1] 不通过。
    if (node.val === node.next.val) {
      node.next = node.next.next;
    } else {
      node = node.next;
    }
  }
  return head;
};
